src/diffusers/pipelines/unidiffuser/modeling_text_decoder.py

```diff
@@ -0,0 +1,220 @@
+
+    def forward(
+        self,
+        tokens: torch.Tensor,
+        prefix: torch.Tensor,
+        mask: Optional[torch.Tensor] = None,
+        labels: Optional[torch.Tensor] = None,
+    ):
+        """
+        Args:
+            tokens (`torch.Tensor` of shape `(N, max_seq_len)`):
+                Text tokens to use for inference.
+            prefix (`torch.Tensor` of shape `(N, prefix_length, 768)`):
+                Prefix embedding to preprend to the embedded tokens.
+            mask (`torch.Tensor` of shape `(N, prefix_length + max_seq_len, 768)`, *optional*):
+                Attention mask for the prefix embedding.
+            labels (`torch.Tensor`, *optional*):
+                TODO
+        """
+        embedding_text = self.transformer.transformer.wte(tokens)
+        hidden = self.encode_prefix(prefix)
+        prefix = self.decode_prefix(hidden)
+        embedding_cat = torch.cat((prefix, embedding_text), dim=1)
+
+        if labels is not None:
+            dummy_token = self.get_dummy_token(tokens.shape[0], tokens.device)
+            labels = torch.cat((dummy_token, tokens), dim=1)
+        out = self.transformer(inputs_embeds=embedding_cat, labels=labels, attention_mask=mask)
+        if self.prefix_hidden_dim is not None:
+            return out, hidden
+        else:
+            return out
+
+    def get_dummy_token(self, batch_size: int, device: torch.device) -> torch.Tensor:
+        return torch.zeros(batch_size, self.prefix_length, dtype=torch.int64, device=device)
+
+    @torch.no_grad()
+    def generate_captions(self, tokenizer, features, device):
+        """
+        Generate captions given text embedding features. Returns list[L].
+
+        Args:
+            tokenizer (`GPT2Tokenizer`):
+                Tokenizer of class
+                [GPT2Tokenizer](https://huggingface.co/docs/transformers/model_doc/gpt2#transformers.GPT2Tokenizer) for
+                tokenizing input to the text decoder model.
+            features (`torch.Tensor` of shape `(B, L, D)`):
+                Text embedding features to generate captions from.
+            device:
+                Device to perform text generation on.
+        """
+
+        features = torch.split(features, 1, dim=0)
+        generated_captions = []
+        for feature in features:
+            feature = self.decode_prefix(feature.to(device))  # back to the clip feature
+            # Only support beam search for now
+            generated_captions.append(self.generate_beam(tokenizer, embed=feature, device=device)[0])
+        return generated_captions
+
+    @torch.no_grad()
+    def generate_beam(
+        self,
+        tokenizer,
+        prompt=None,
+        embed=None,
+        device=None,
+        beam_size: int = 5,
+        entry_length: int = 67,
+        temperature: float = 1.0,
+        stop_token: str = "<|EOS|>",
+    ):
+        """
+        Generates text using the given tokenizer and text prompt or token embedding via beam search.
```

Comment:
Help me understand this a bit. Why would there be a need to generate text from a given text prompt?

Reply:
Sorry, I think I wrote this docstring in a confusing way. In the context of UniDiffuser sampling, we use this function to generate output text (when appropriate) from the text latents after we process the CLIP-embedded input prompt using the `unet` (`UniDiffuserModel`) model. The method accepts both `prompt` and `embed` arguments, for input tokens and embeddings respectively, but we only ever call it with input embeddings (as described above): https://github.com/huggingface/diffusers/blob/2fdb7b871b994e96507d7d71debc38b47ebd4c99/src/diffusers/pipelines/unidiffuser/modeling_text_decoder.py#L135

Reply:
Oh okay. Yeah, then I guess we need to it make it a bit clearer from the code?